TrustedTypes: WorkerGlobalScope: importScripts() (#42396)
mdn/content: files/en-us/web/api/workerglobalscope/importscripts/index.md

{{APIRef("Web Workers API")}}{{AvailableInWorkers("worker")}}

> [!WARNING]

> > The parameters passed to this method represent the URLs of classic scripts to be imported into a worker.

Reviewer: These docs talk a lot about "classic" and "module" scripts. Would be great to have somewhere to link where we can explain them. Glossary page?

### Exceptions

- `NetworkError`
  - : Imported scripts were served without a `text/javascript` media (MIME) type or without one of the permitted [legacy JavaScript MIME types](/en-US/docs/Web/HTTP/Guides/MIME_types#legacy_javascript_mime_types).
- {{jsxref("TypeError")}}
  - : Thrown if the current {{domxref("WorkerGlobalScope")}} is a module (use [`import`](/en-US/docs/Web/JavaScript/Reference/Statements/import) instead).
    It may also happen if any parameter is not a {{domxref("TrustedScriptURL")}} and the site is [enforcing trusted types](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types).

## Description

The **`importScripts()`** method synchronously imports one or more scripts into the scope of a classic worker.

Unlike the initial classic module script, which must be same-origin with its document, this method can import cross-origin scripts unless blocked by a {{httpheader("Cross-Origin-Resource-Policy")}}, [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/Guides/CSP), or some other security mechanism.

> Note that because classic scripts are fetched in `no-cors` mode, they are not blocked by CORS.

Reviewer: CORS does not block anything. I think it's important to be precise here because everyone treats CORS as a mechanism that prevents things, and it's the opposite, and we shouldn't perpetuate that. AFAIK the only place where we explain `no-cors` is https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#making_cross-origin_requests, but that's in the context of `fetch()` so it's a little different (because it expects the caller will usually want to read the result, which is not the case here).

```suggestion
Note that because classic scripts are fetched in `no-cors` mode, they can be fetched cross-origin even if the server does not set the appropriate CORS headers.
```

### Security considerations

The parameters specify scripts to be imported into the scope of a classic worker.
If the input is provided by a user, this is a possible vector for [cross-site scripting (XSS)](/en-US/docs/Web/Security/Attacks/XSS) attacks.

It is extremely risky to accept and execute arbitrary URLs from untrusted origins.

> A website should control what scripts that are allowed to run using a [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/Guides/CSP) with the [`script-src`](/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/worker-src) directive (or a fallback defined in [`default-src`](/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/default-src)).

Author: Note, wierd as it is, `script-src`, not `worker-src` for classic workers.

Trusted types are not yet supported on all browsers, so first we define the [trusted types tinyfill](/en-US/docs/Web/API/Trusted_Types_API#trusted_types_tinyfill).
This acts as a transparent replacement for the trusted types JavaScript API:

```js
if (typeof trustedTypes === "undefined")
  trustedTypes = { createPolicy: (n, rules) => rules };
```

Next we create a {{domxref("TrustedTypePolicy")}} that defines a {{domxref("TrustedTypePolicy/createScriptURL", "createScriptURL()")}} method for transforming input strings into {{domxref("TrustedScriptURL")}} instances.

For the purpose of this example we'll assume that we want to allow a predefined set of URLs in the `scriptAllowList` array and log any other scripts.

```js
const scriptAllowList = [
  // Some list of allowed URLs
];
const policy = trustedTypes.createPolicy("script-url-policy", {
  createScriptURL(input) {
    if (scriptAllowList.includes(input)) {
      return input; // allow the script
    }
    console.log(`Script not in scriptAllowList: ${input}`);
    return ""; // Block the script
  },
});
```

> Then we use the `policy` object to create a `trustedScript` object from a potentially unsafe input string:

Reviewer: ```suggestion
Then we use the `policy` object to create a `TrustedScript` object from a potentially unsafe input string:
```

```js
// The potentially malicious string
// We won't be including untrustedScript in our scriptAllowList array
const untrustedScript = "https://evil.example.com/import_worker.js";

// Create a TrustedScriptURL instance using the policy
const trustedScriptURL = policy.createScriptURL(untrustedScript);
```

> The `trustedScriptURL` property can now be used when importing the script in a classic worker:

Reviewer: ```suggestion
The `TrustedScriptURL` object can now be used when importing the script in a classic worker:
```
